src/tree/remoteProject/RemoteFunctionsTreeItem.ts

```diff
@@ -42,28 +41,35 @@ export class RemoteFunctionsTreeItem extends FunctionsTreeItemBase {
             this._nextLink = undefined;
         }
 
+        /*
+            Related to issue: https://github.com/microsoft/vscode-azurefunctions/issues/3179
+            Sometimes receive a 'BadGateway' error on initial fetch, but consecutive re-fetching usually fixes the issue.
+            Under these circumstances, we will attempt to do the call 3 times during warmup before throwing the error.
+        */
+        const retries = 3;
         const client = await this.parent.site.createClient(context);
-        let funcs: FunctionEnvelope[];
-        const maxTime = Date.now() + 2 * 60 * 1000;
-        let attempt = 1;
-        while (true) {
-            // Load more currently broken https://github.com/Azure/azure-sdk-for-js/issues/20380
-            funcs = await client.listFunctions();
 
-            // https://github.com/Azure/azure-functions-host/issues/3502
-            if (!isArray(funcs)) {
-                throw new Error(localize('failedToList', 'Failed to list functions.'));
-            }
+        const funcs = await retry<FunctionEnvelope[]>(
+            async (attempt: number) => {
+                // Load more currently broken https://github.com/Azure/azure-sdk-for-js/issues/20380
+                const response = await client.listFunctions();
 
-            // Retry listing functions if all we see is a "WarmUp" function, an internal function that goes away once the app is ...warmed up
-            if (Date.now() > maxTime || !(funcs.length === 1 && isWarmupFunction(funcs[0]))) {
-                context.telemetry.measurements.listFunctionsAttempt = attempt;
-                break;
-            } else {
-                attempt += 1;
-                await delay(10 * 1000);
-            }
-        }
+                // https://github.com/Azure/azure-functions-host/issues/3502
+                if (!Array.isArray(response)) {
+                    throw new Error(localize('failedToList', 'Failed to list functions.'));
+                }
+
+                // Retry listing functions if all we see is a "WarmUp" function, an internal function that goes away once the app is ...warmed up
+                if (!(response.length === 1 && isWarmupFunction(response[0]))) {
+                    context.telemetry.measurements.listFunctionsAttempt = attempt;
+                } else {
+                    throw new Error(localize('warmUpFunction', 'Server response was a WarmUp function.'));
```

Comment:
I think we should just stick with saying `Failed to list functions` rather than specifying the server response.

Though isn't the condition for the else clause in this case that it's _not_ a warmup function?

Reply:
It looks like it will only pass the final check if it's not a warm-up function.  If it is a warm-up function it will throw an error so that we can attempt again.  I'm cool with changing the error messages to be the same